keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -60,35 +62,43 @@ class SqlArtifactRepository(
       .and(DELIVERY_ARTIFACT.TYPE.eq(type.name))
       .fetchOne() != null
 
-  override fun versions(artifact: DeliveryArtifact): List<String> =
-    if (isRegistered(artifact.name, artifact.type)) {
+  override fun versions(artifact: DeliveryArtifact, statuses: List<ArtifactStatus>): List<String> {
+    val status = statuses.map { it.toString() }
+    return if (isRegistered(artifact.name, artifact.type)) {
       jooq
         .select(DELIVERY_ARTIFACT_VERSION.VERSION)
         .from(DELIVERY_ARTIFACT, DELIVERY_ARTIFACT_VERSION)
         .where(DELIVERY_ARTIFACT.UID.eq(DELIVERY_ARTIFACT_VERSION.DELIVERY_ARTIFACT_UID))
         .and(DELIVERY_ARTIFACT.NAME.eq(artifact.name))
         .and(DELIVERY_ARTIFACT.TYPE.eq(artifact.type.name))
+        .and(DELIVERY_ARTIFACT_VERSION.STATUS.`in`(*status.toTypedArray()))
         .fetch()
         .getValues(DELIVERY_ARTIFACT_VERSION.VERSION)
         .sortAppVersion()
     } else {
       throw NoSuchArtifactException(artifact)
     }
+  }
 
   override fun latestVersionApprovedIn(
     deliveryConfig: DeliveryConfig,
     artifact: DeliveryArtifact,
-    targetEnvironment: String
+    targetEnvironment: String,
+    statuses: List<ArtifactStatus>
   ): String? {
     val environment = deliveryConfig.environmentNamed(targetEnvironment)
-    return jooq
+    val approvedVersions = jooq
       .select(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION)
       .from(ENVIRONMENT_ARTIFACT_VERSIONS)
       .where(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(deliveryConfig.getUidFor(environment)))
       .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID.eq(artifact.uid))
       .orderBy(ENVIRONMENT_ARTIFACT_VERSIONS.APPROVED_AT.desc())
-      .limit(1)
-      .fetchOne(0, String::class.java)
+      .fetch()
+      .getValues(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION)
+    val versionsWithCorrectStatus = versions(artifact, statuses)
+
+    // return the latest version that has been approved with the correct status
+    return approvedVersions.intersect(versionsWithCorrectStatus).firstOrNull()
```

Comment:
Is there a reason we can't just add one more table join to the query rather than having to do this with the intersect of 2 different query results?

Reply:
I consulted with @asher and he advised doing it this way. Also, my sql is not that great, and I found the syntax for joining to be extremely confusing.

Reply:
I think I was thinking more in terms of multiple queries being ok and sometimes preferred where everything fetched is needed. The intersect is a sign of fetching more from the DB than we need which is better to avoid. For joins, I prefer to be explicit about the join type and related columns, makes it easier to understand index needs - I'd have to look at `EXPLAIN` output to check if the query in `versions()` is efficient. I need to get up to speed on how the schema has evolved around environments, I'm not currently sure how `ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION` differs from `DELIVERY_ARTIFACT_VERSION.VERSION` so I'm hesitant to write an example query here but would be happy to work it out in intellj w/you along with how to validate index usage and efficiency.

Reply:
Those two things hold the same value. `DELIVERY_ARTIFACT_VERSION` holds all known versions per artifact. `ENVIRONMENT_ARTIFACT_VERSIONS` holds the versions that have been approved and/or deployed to a given environment.

Reply:
I made it bester